frontend/src/pages/question/questionBuilder/QueryJsonEditor.jsx

```diff
@@ -0,0 +1,266 @@
+      }
+    }
+
+    setErrorMessages(errMessage);
+  }
+
+  function updateJson(e) {
+    // updated_src is the updated graph RJV gives back
+    const data = e.updated_src;
+    validateQueryGraph(data);
+    updateQueryGraph(data);
+  }
+
+  function onUpload(acceptedFiles) {
+    acceptedFiles.forEach((file) => {
+      const fr = new window.FileReader();
+      fr.onloadstart = () => pageStatus.setLoading('Loading Query Graph');
+      fr.onload = (e) => {
+        const fileContents = e.target.result;
+        try {
+          let graph = JSON.parse(fileContents);
+
+          // rip out just the query graph and convert to our internal
+          // representation
+          if (graph.query_graph) {
+            graph = graph.query_graph;
+          } else if (graph.question_graph) {
+            graph = graph.question_graph;
+          }
+          if (graph.nodes && Array.isArray(graph.nodes)) {
+            graph = queryGraphUtils.convert.reasonerToInternal(graph);
+          }
+          validateQueryGraph(graph);
+          updateQueryGraph(graph);
+          initialQueryGraph.current = questionStore.query_graph;
+          pageStatus.setSuccess();
+        } catch (err) {
+          pageStatus.setFailure('Failed to read this query graph. Are you sure this is valid json?');
+        }
+      };
+      fr.onerror = () => {
+        displayAlert('error', 'Sorry but there was a problem uploading the file. The file may be invalid JSON.');
+      };
+      fr.readAsText(file);
+    });
+  }
+
+  useEffect(() => {
+    validateQueryGraph(questionStore.query_graph);
+    updateQueryGraph(questionStore.query_graph);
+    initialQueryGraph.current = questionStore.query_graph;
+  }, [questionStore.query_graph]);
+
+  const query_graph_list_format = useMemo(() => {
+    if (!errorMessages.length) {
+      return queryGraphUtils.convert.internalToReasoner(queryGraph);
+    }
+    return queryGraphUtils.getEmptyGraph();
+  }, [queryGraph]);
+
+  return (
+    <Dialog
+      open={show}
+      fullWidth
+      maxWidth="lg"
+    >
+      <DialogTitle style={{ padding: 0 }}>
+        <div id="jsonEditorTitle">
+          <div>
+            <IconButton
+              style={{ fontSize: '18px' }}
+              title="Revert"
+              disabled={!pageStatus.displayPage}
+              onClick={() => {
+                validateQueryGraph(initialQueryGraph.current);
+                updateQueryGraph(initialQueryGraph.current);
+              }}
+            >
+              <UndoIcon />
+            </IconButton>
+            <label htmlFor="jsonEditorUpload" id="uploadIconLabel">
+              <input
+                accept=".json"
+                style={{ display: 'none' }}
+                type="file"
+                id="jsonEditorUpload"
+                onChange={(e) => onUpload(e.target.files)}
+                disabled={!pageStatus.displayPage}
```

Comment:
These buttons are disabled when an invalid question is uploaded and the state persists even when closing and re-opening the JSON editor. Maybe we should keep them enabled to let users re-try the upload.

Reply:
Good catch.